File: src-tauri/src/command_contract.rs
use serde_json::Value;
use std::sync::OnceLock;

const CONTRACT_JSON: &str = include_str!("../../contracts/command_contract.json");

fn table() -> &'static Value {
    static TABLE: OnceLock<Value> = OnceLock::new();
    TABLE.get_or_init(|| {
        serde_json::from_str(CONTRACT_JSON).expect("embedded command contract must be valid JSON")
    })
}

fn resolve_ref<'a>(root: &'a Value, reference: &str) -> Result<&'a Value, String> {
    let name = reference
        .strip_prefix("#/definitions/")
        .ok_or_else(|| format!("contract contains unsupported ref: {reference}"))?;
    root.get("definitions")
        .and_then(|definitions| definitions.get(name))
        .ok_or_else(|| format!("contract contains unknown ref: {reference}"))
}

fn matches(value: &Value, schema: &Value, root: &Value, path: &str) -> bool {
    validate(value, schema, root, path).is_ok()
}

fn validate(value: &Value, schema: &Value, root: &Value, path: &str) -> Result<(), String> {
    if let Some(reference) = schema.get("$ref").and_then(Value::as_str) {
        return validate(value, resolve_ref(root, reference)?, root, path);
    }
    if let Some(candidates) = schema.get("oneOf").and_then(Value::as_array) {
        let count = candidates
            .iter()
            .filter(|candidate| matches(value, candidate, root, path))
            .count();
        return if count == 1 {
            Ok(())
        } else {
            Err(format!("{path} must match exactly one allowed shape"))
        };
    }
    if let Some(expected) = schema.get("const") {
        if value != expected {
            return Err(format!("{path} does not match the required constant"));
        }
    }
    if let Some(allowed) = schema.get("enum").and_then(Value::as_array) {
        if !allowed.contains(value) {
            return Err(format!("{path} is not an allowed value"));
        }
    }

    let Some(kind) = schema.get("type").and_then(Value::as_str) else {
        return Ok(());
    };
    match kind {
        "object" => {
            let object = value
                .as_object()
                .ok_or_else(|| format!("{path} must be object"))?;
            if let Some(maximum) = schema.get("maxProperties").and_then(Value::as_u64) {
                if object.len() as u64 > maximum {
                    return Err(format!("{path} has too many properties"));
                }
            }
            if let Some(required) = schema.get("required").and_then(Value::as_array) {
                for key in required.iter().filter_map(Value::as_str) {
                    if !object.contains_key(key) {
                        return Err(format!("{path}.{key} is required"));
                    }
                }
            }
            if let Some(properties) = schema.get("properties").and_then(Value::as_object) {
                for (key, property_schema) in properties {
                    if let Some(property) = object.get(key) {
                        validate(property, property_schema, root, &format!("{path}.{key}"))?;
                    }
                }
            }
        }
        "array" => {
            let array = value
                .as_array()
                .ok_or_else(|| format!("{path} must be array"))?;
            if let Some(item_schema) = schema.get("items") {
                for (index, item) in array.iter().enumerate() {
                    validate(item, item_schema, root, &format!("{path}[{index}]"))?;
                }
            }
        }
        "string" if !value.is_string() => return Err(format!("{path} must be string")),
        "boolean" if !value.is_boolean() => return Err(format!("{path} must be boolean")),
        "null" if !value.is_null() => return Err(format!("{path} must be null")),
        "number" if !value.is_number() => return Err(format!("{path} must be number")),
        "integer" if value.as_u64().is_none() && value.as_i64().is_none() => {
            return Err(format!("{path} must be integer"));
        }
        _ => {}
    }

    if value.is_number() {
        let number = value
            .as_f64()
            .ok_or_else(|| format!("{path} must be a finite number"))?;
        if schema
            .get("minimum")
            .and_then(Value::as_f64)
            .is_some_and(|minimum| number < minimum)
        {
            return Err(format!("{path} is below minimum"));
        }
        if schema
            .get("maximum")
            .and_then(Value::as_f64)
            .is_some_and(|maximum| number > maximum)
        {
            return Err(format!("{path} is above maximum"));
        }
    }
    Ok(())
}

fn find_kind(collection: &str, kind: &str) -> Option<&'static Value> {
    table()
        .get(collection)?
        .as_array()?
        .iter()
        .find(|entry| entry.get("kind").and_then(Value::as_str) == Some(kind))
}

pub fn is_error_code(code: &str) -> bool {
    table()["errorCodes"]
        .as_array()
        .is_some_and(|codes| codes.iter().any(|entry| entry["code"] == code))
}

pub fn error_codes() -> Vec<&'static str> {
    table()["errorCodes"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|entry| entry["code"].as_str())
        .collect()
}

pub fn is_event_kind(kind: &str) -> bool {
    find_kind("events", kind).is_some()
}

pub fn validate_command_payload(kind: &str, payload: &Value) -> Result<(), String> {
    let command =
        find_kind("commands", kind).ok_or_else(|| format!("unknown command kind: {kind}"))?;
    validate(payload, &command["payload"], table(), "payload")
}

File: engine/src/command_contract.cpp
#include "command_contract.hpp"

#include "command_contract_embed.hpp"

#include <cmath>
#include <cstddef>
#include <string_view>

namespace rmx::contract {

namespace {

[[noreturn]] void invalid(const std::string& path, const std::string& reason) {
    throw ValidationError(path + " " + reason);
}

const nlohmann::json& resolve_ref(const nlohmann::json& root, const std::string& ref) {
    constexpr std::string_view kPrefix = "#/definitions/";
    if (!ref.starts_with(kPrefix)) invalid("contract", "contains unsupported ref: " + ref);
    const auto name = ref.substr(kPrefix.size());
    const auto& definitions = root.at("definitions");
    if (!definitions.contains(name)) invalid("contract", "contains unknown ref: " + ref);
    return definitions.at(name);
}

void validate(const nlohmann::json& value, const nlohmann::json& schema,
              const nlohmann::json& root, const std::string& path);

bool matches(const nlohmann::json& value, const nlohmann::json& schema,
             const nlohmann::json& root, const std::string& path) {
    try {
        validate(value, schema, root, path);
        return true;
    } catch (const ValidationError&) {
        return false;
    }
}

void validate_number_bounds(const nlohmann::json& value, const nlohmann::json& schema,
                            const std::string& path) {
    if (!value.is_number()) return;
    const auto number = value.get<double>();
    if (!std::isfinite(number)) invalid(path, "must be finite");
    if (schema.contains("minimum") && number < schema.at("minimum").get<double>())
        invalid(path, "is below minimum");
    if (schema.contains("maximum") && number > schema.at("maximum").get<double>())
        invalid(path, "is above maximum");
}

void validate(const nlohmann::json& value, const nlohmann::json& schema,
              const nlohmann::json& root, const std::string& path) {
    if (schema.contains("$ref")) {
        validate(value, resolve_ref(root, schema.at("$ref").get<std::string>()), root, path);
        return;
    }
    if (schema.contains("oneOf")) {
        std::size_t count = 0;
        for (const auto& candidate : schema.at("oneOf"))
            if (matches(value, candidate, root, path)) ++count;
        if (count != 1) invalid(path, "must match exactly one allowed shape");
        return;
    }
    if (schema.contains("const") && value != schema.at("const"))
        invalid(path, "does not match the required constant");
    if (schema.contains("enum")) {
        bool found = false;
        for (const auto& candidate : schema.at("enum")) found = found || value == candidate;
        if (!found) invalid(path, "is not an allowed value");
    }

    if (!schema.contains("type")) return;
    const auto type = schema.at("type").get<std::string>();
    if (type == "object") {
        if (!value.is_object()) invalid(path, "must be object");
        if (schema.contains("maxProperties") &&
            value.size() > schema.at("maxProperties").get<std::size_t>())
            invalid(path, "has too many properties");
        if (schema.contains("required")) {
            for (const auto& key_json : schema.at("required")) {
                const auto key = key_json.get<std::string>();
                if (!value.contains(key)) invalid(path + "." + key, "is required");
            }
        }
        if (schema.contains("properties")) {
            for (auto it = schema.at("properties").begin(); it != schema.at("properties").end();
                 ++it) {
                if (value.contains(it.key()))
                    validate(value.at(it.key()), it.value(), root, path + "." + it.key());
            }
        }
        return;
    }
    if (type == "array") {
        if (!value.is_array()) invalid(path, "must be array");
        if (schema.contains("items")) {
            for (std::size_t i = 0; i < value.size(); ++i)
                validate(value.at(i), schema.at("items"), root,
                         path + "[" + std::to_string(i) + "]");
        }
        return;
    }
    if (type == "string" && !value.is_string()) invalid(path, "must be string");
    if (type == "boolean" && !value.is_boolean()) invalid(path, "must be boolean");
    if (type == "null" && !value.is_null()) invalid(path, "must be null");
    if (type == "number" && !value.is_number()) invalid(path, "must be number");
    if (type == "integer" && !value.is_number_integer() && !value.is_number_unsigned())
        invalid(path, "must be integer");
    validate_number_bounds(value, schema, path);
}

const nlohmann::json* find_kind(const char* collection, const std::string& kind) {
    for (const auto& entry : table().at(collection))
        if (entry.at("kind") == kind) return &entry;
    return nullptr;
}

void validate_command_member(const std::string& kind, const nlohmann::json& value,
                             const char* schema_key) {
    const auto* command = find_kind("commands", kind);
    if (command == nullptr) invalid("kind", "is unknown: " + kind);
    validate(value, command->at(schema_key), table(), schema_key);
}

}  // namespace

const nlohmann::json& table() {
    static const nlohmann::json value = [] {
        auto parsed = nlohmann::json::parse(kCommandContractJson, nullptr, false);
        if (parsed.is_discarded() || !parsed.is_object())
            throw std::runtime_error("embedded command contract is invalid JSON");
        return parsed;
    }();
    return value;
}

bool is_command_kind(const std::string& kind) { return find_kind("commands", kind) != nullptr; }

bool is_error_code(const std::string& code) {
    for (const auto& entry : table().at("errorCodes"))
        if (entry.at("code") == code) return true;
    return false;
}

bool is_event_kind(const std::string& kind) { return find_kind("events", kind) != nullptr; }

// 指令錯誤回覆的執法表:code 必須屬於該指令宣告的 errors 清單。
// 未知 kind(contract 之外,如 dispatch 前 unsupported_version 路徑)= 跳過。
bool is_declared_error(const std::string& kind, const std::string& code) {
    const auto* command = find_kind("commands", kind);
    if (command == nullptr) return true;
    for (const auto& declared : command->at("errors"))
        if (declared == code) return true;
    return false;
}

void validate_command_payload(const std::string& kind, const nlohmann::json& payload) {
    validate_command_member(kind, payload, "payload");
}

void validate_result(const std::string& kind, const nlohmann::json& result) {
    validate_command_member(kind, result, "result");
}

}  // namespace rmx::contract
